### PR TITLE
Use Azure upload method instead of our own implementation

### PR DESCRIPTION
**Pull request by dadoonet**
We are not following the Azure documentation about uploading blobs to Azure storage. https://docs.microsoft.com/en-us/azure/storage/blobs/storage-java-how-to-use-blob-storage#upload-a-blob-into-a-container

Instead we are using our own implementation which might cause some troubles and rarely some blobs can be not immediately commited just after we close the stream. Using the standard implementation provided by Azure team should allow us to benefit from all the magic Azure SDK team already wrote.

And well... Let's just read the doc!

### CODE REVIEW

**Review by imotov — APPROVED** [review]
LGTM. Thanks for finding it out! Can we backport it? I think it would be great to switch to this sooner rather then later. It looks like we are missing quite a bit of logic by uploading the blob our old way.

**Comment by dadoonet** [comment]
@imotov Thanks for the review. I did some manual testings this morning and it does not work.

Apparently the file `master.dat-temp` is not written in the azure container... 
Getting an exception saying that the container does not exist although I can see it in the azure Web interface... 

I'm digging... Probably something stupid on my end. :)

**Comment by dadoonet** [comment]
@imotov I worked on IT so we can now pass them when needed (still a manual operation).
I tried to simplify and remove non needed things.

I tested everything manually:

* Install elasticsearch 7.0.0-alpha1-SNAPSHOT
* Install repository-azure plugin
* Run the following test:

```sh
# Clean test env
curl -XDELETE localhost:9200/foo?pretty
curl -XDELETE localhost:9200/_snapshot/my_backup1/snap1?pretty
curl -XDELETE localhost:9200/_snapshot/my_backup1?pretty

# Create data
curl -XPUT localhost:9200/foo/doc/1?pretty -H 'Content-Type: application/json' -d '{
 "foo": "bar"
}'
curl -XPOST localhost:9200/foo/_refresh?pretty

# Create repository using default account
curl -XPUT localhost:9200/_snapshot/my_backup1?pretty -H 'Content-Type: application/json' -d '{
 "type": "azure"
}'

# Backup
curl -XPOST "localhost:9200/_snapshot/my_backup1/snap1?pretty&wait_for_completion=true"

# Delete existing index
curl -XDELETE localhost:9200/foo?pretty

# Restore using default account
curl -XPOST "localhost:9200/_snapshot/my_backup1/snap1/_restore?pretty&wait_for_completion=true"

# Check
curl -XGET localhost:9200/foo/_search?pretty

# Remove backup
curl -XDELETE localhost:9200/_snapshot/my_backup1/snap1?pretty
curl -XDELETE localhost:9200/_snapshot/my_backup1?pretty
```

Everything is correct. I'm going to test with a bigger dataset now and check everything works.
Could you give a final review on the code please as I changed some code recently?

Thanks!

**Comment by dadoonet** [comment]
I tested with much more data (300mb) and everything is working well.
LMK! :)

**Comment by imotov** [comment]
@dadoonet would it make sense to base this tests on [`ESBlobStoreRepositoryIntegTestCase`](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESBlobStoreRepositoryIntegTestCase.java)? I think this base class has most of the tests that we want to run a repo to ensure that it behaves reasonably. If you find it lacking something, I think it would make sense to extend it so all other repos would benefit

**Comment by dadoonet** [comment]
@imotov Great! I did not remember about that class. Yeah. Definitely better using it as well.

I pushed new changes.

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureSnapshotRestoreTests.java [hunk 1]
```diff
@@ -36,49 +36,59 @@
...
+import org.elasticsearch.test.ESIntegTestCase.ThirdParty;
 import org.elasticsearch.test.store.MockFSDirectoryService;
 import org.elasticsearch.test.store.MockFSIndexStore;
 import org.junit.After;
-import org.junit.Before;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
 
 import java.net.URISyntaxException;
+import java.net.UnknownHostException;
```
Comment (imotov) [inline plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureSnapshotRestoreTests.java]: Unused.

**Review by imotov — APPROVED** [review]
LGTM. Thanks for moving the tests! Now we just need to figure out how to get it into CI to run on a regular basis.

**Comment by dadoonet** [comment]
I backported it on 6.x yet.

I'm planning to backport on 6.0 but it's a bit harder as some PR have not been merged to 6.0 like #23518 and #23405.

**Comment by dadoonet** [comment]
Backported to 6.0 as well with 9aa5595d199d41f7681d6814616dd73d52a61b66

**Comment by dadoonet** [comment]
Backported to 5.6 with https://github.com/elastic/elasticsearch/pull/26839/commits/28f17a72f617bde54ee6e1071e1491e03740d967 (see #26839)

**Pull request merged**